ext/standard/streamsfuncs.c

```diff
@@ -756,116 +809,141 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
 /* }}} */
 
 /* {{{ Runs the select() system call on the sets of streams with a timeout specified by tv_sec and tv_usec */
-PHP_FUNCTION(stream_select)
-{
-	zval *r_array, *w_array, *e_array;
-	struct timeval tv, *tv_p = NULL;
-	fd_set rfds, wfds, efds;
-	php_socket_t max_fd = 0;
-	int retval, sets = 0;
-	zend_long sec, usec = 0;
-	bool secnull;
-	bool usecnull = 1;
-	int set_count, max_set_count = 0;
-
-	ZEND_PARSE_PARAMETERS_START(4, 5)
-		Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
-		Z_PARAM_LONG_OR_NULL(sec, secnull)
-		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG_OR_NULL(usec, usecnull)
-	ZEND_PARSE_PARAMETERS_END();
-
-	FD_ZERO(&rfds);
-	FD_ZERO(&wfds);
-	FD_ZERO(&efds);
-
-	if (r_array != NULL) {
-		set_count = stream_array_to_fd_set(r_array, &rfds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+PHP_FUNCTION(stream_select) {
+    zval *r_array, *w_array, *e_array;
+    struct timeval tv, *tv_p = NULL;
+    fd_bigset rfds, wfds, efds;
+    php_socket_t max_fd = -1;
+    int retval = 0;
+    zend_long sec, usec = 0;
+    bool secnull;
+    bool usecnull = 1;
+
+    ZEND_PARSE_PARAMETERS_START(4, 5)
+        Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
+        Z_PARAM_LONG_OR_NULL(sec, secnull)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG_OR_NULL(usec, usecnull)
+    ZEND_PARSE_PARAMETERS_END();
+
+    // Determine the largest file descriptor
+    if (r_array != NULL) {
+        max_fd = stream_array_get_max_fd(r_array);
+    }
+    if (w_array != NULL) {
+        php_socket_t tmp_max_fd = stream_array_get_max_fd(w_array);
+        if (tmp_max_fd > max_fd) {
+            max_fd = tmp_max_fd;
+        }
+    }
+    if (e_array != NULL) {
+        php_socket_t tmp_max_fd = stream_array_get_max_fd(e_array);
+        if (tmp_max_fd > max_fd) {
+            max_fd = tmp_max_fd;
+        }
+    }
 
-	if (w_array != NULL) {
-		set_count = stream_array_to_fd_set(w_array, &wfds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+    if (max_fd == -1) {
+        zend_value_error("No stream arrays were passed or all arrays were empty");
+        RETURN_THROWS();
+    }
 
-	if (e_array != NULL) {
-		set_count = stream_array_to_fd_set(e_array, &efds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+    // Allocate fd_bigset based on max_fd
+    FD_BIGSET_ZERO(&rfds, max_fd + 1);
+    FD_BIGSET_ZERO(&wfds, max_fd + 1);
+    FD_BIGSET_ZERO(&efds, max_fd + 1);
```

Comment:
If I'm understanding correctly, `stream_array_get_max_fd` will return `0` with empty arrays, resulting in an attempt to allocate 1*3 byte here and continue execution. It's probably not a big deal, but perhaps if there are no streams, we can bail?

Reply:
I think lines 848-851 handles that.

Reply:
It handles the case where the max is -1, not zero. The max fd function returns 0 for an empty array.

Reply:
I decided not to use stream_array_get_max_fd, since it involves casting the stream resources, and during testing it reveals that for streamWrapper implementations not implementing the cast method, warnings are raised. To avoid casting stream resources twice (first for counting to decide the size of the bitmap, then again to build the bitmap), I simply allocate a bitmap that is large enough when `sysconf(_SC_OPEN_MAX)` exists, and a bitmap sized equal to FD_SETSIZE if not. Since the memory is allocated once and immediately released when the function returns it won't affect memory usage of the applicatin - but it will temporarily require up to a few 100 KB at most if the machine is configured to allow a very large number of file descriptors in processes. My machine is configured with an ulimit of 1 million which works fine.